### DISCUSSION TITLE
[WS2122] Homework 1 Discussion

### DISCUSSION THREAD

**Discussion by changkun** [mimuc/gp-ws2021#29] [Discussion] [answered]
This thread is open for the discussion of homework 1.

**Comment by Consti10** [comment]
Hello, quick Question - What should you assign the "idx" members of class Vertex,Halfedge,Edge and Face ?
Is it just a value increasing with the n of (Vertices,Halfedges,...) , e.g. first Vertex->idx 0, second Vertex->idx 1, ... ?

**Reply by changkun** [reply]
In my solution, they are irrelevant to the construction process actually. However, they are useful for later homeworks. Perhaps that's a little bit misleading in this homework because they are not really used. Let's ignore them :)

If you interested to initialize them, you may just loop what we collected in the array then assign an incremental ID to them at the bottom of the entire implementation of the function, for instance:

```
    // reset indices
    let index = 0;
    this.verts.forEach(v => {
      v.idx = index++;
    });
    index = 0;
    this.edges.forEach(e => {
      e.idx = index++;
    });
    index = 0;
    this.faces.forEach(f => {
      f.idx = index++;
    });
    index = 0;
    this.halfedges.forEach(h => {
      h.idx = index++;
    });
```

**Reply by Consti10** [reply]
okay, thanks. I am still in the earliest process, so I'l just leave them at -1 for now, if the rendering isn't affected by them anyways.

**Reply by Consti10** [reply]
Does the same hold true for onBoundary (boolean) value of Halfedge ?

**Reply by changkun** [reply]
onBoundary is tricky. It is not affected by rendering but will be useful later. It might be worth thinking about how to record this information while construction, so that we don't have to check either an edge is on the boundary at runtime.

**Reply by Consti10** [reply]
Is "onBoundary==true" equivalent to "member twin of halfedge==null" ?

**Reply by changkun** [reply]
I think they are not equivalent. If an edge is on the boundary, you may still have to create the two halfedges to reserve the ability of traversal. For instance, you may want to travel through all edges of a boundary.

**Reply by Consti10** [reply]
I mistook boundary and edge there.

**Comment by Consti10** [comment]
For rendering a single triangle, is the following calculation correct ?
You need
Vertices   : 3
Halfedges: 3*2=6 
Edges      : 3
Faces       : 1

**Reply by Consti10** [reply]
![Triangle](https://user-images.githubusercontent.com/16462285/137888836-2ebcc8e7-c7ba-42ab-a328-792b511e82e3.jpg)

**Reply by changkun** [reply]
That's correct :)

**Reply by Consti10** [reply]
I think for a single triangle, you actually only need 3 halfedges stored in memory - 1 for each edge. At least I get "some image" this way ;)

**Comment by Consti10** [comment]
Work flow: I found out that one doesn't have to kill and restart with "npm start" to see your code changes on the screen. Just press "ctr+s" to save your changes, then something recompiles the code and you can see your changes in the browser tab.

**Reply by changkun** [reply]
That's right, and it is powered by the webpack.

**Comment by Consti10** [comment]
For discussion: While an algorithm that creates as many vertices as indices creates an Image that looks like the "expect.png" (I think this describes what I do currently best, and I get exactly the bunny) the created data structure is then not "correct", 

I think the right solution should have exactly as many Vertices as positions are supplied, right ?

**Reply by changkun** [reply]
Yes. Rendering the correct picture does not really mean the implementation is correct. But I think if we try to implement the structure the halfedge way, then the implementation should match what we expected.

For solid testings, there are multiple ways. We talked about the Euler formula on the slides, which may be a way for a quick test, although they do not guarantee it is indeed correct. Nevertheless, writing a proper unit test for these types of tasks is also a challenging task, which would also be an interesting thing to see if you find more time to work out a solution :)

**Reply by Consti10** [reply]
Are these 2 assumptions correct ?
Rn for example in my case, assumption 1 fails but assumption 2 holds true:

```
validateUsingEuler(){
    // total number of half-edges must be exactly 2x number of edges
    if(this.halfedges.length!=this.edges.length*2!){
      console.log("n of half-edges and edges*2 does not match"+this.halfedges.length+"!="+(this.edges.length*2));
    }
    // total number of half-edges must be exactly 3x number of faces
    if(this.halfedges.length!=this.faces.length*3){
      console.log("n of half-edges and faces*3 does not macth"+this.halfedges.length+"!="+(this.faces.length*3));
    }
    //console.log("Hmm");
  }
```

**Comment by Consti10** [comment]
I have a problem with the following (example) scenario:
Assume there are
 4 vertices (v0,v1,v2,v3) and 6 indices (0,1,2, 0,3,1)
 They form a quadliteral with 6 half-edges (he1,he2,he3, he4,he5,he6) looking like this:
 
![quadliteral](https://user-images.githubusercontent.com/16462285/137941628-5b72a853-4674-417a-8415-eafc0a3d7d0f.jpg)

 Now the problem is: Vertex 0 (v0) has two out-going halfedges (he1,he4). But One Vertex can only store one Halfedge as a member.

**Reply by JohannesMerkt** [reply]
I believe it doesn't matter which halfedge you are using in the end. As long as the mesh is not non-manifold you should be able to get all other outgoing halfedges with the one being provided.

**Comment by Consti10** [comment]
I have one more scenario I am uncertain about -
If you search the internet for halfedge mesh (images), you always see the left case (same as above, the edge is formed by 2 halfedges pointing in different directions.

However, depending on how the indices are used, the right scenario is possible, too - there the edge is formed by 2 halfedges pointing
in the same direction. But they are not the same, since "next" points to a different halfedge.

![quadliteral_weird](https://user-images.githubusercontent.com/16462285/138144764-1bf56cfb-9e15-4d91-9ee5-906a34512af0.jpg)

Are these both valid scenarios ? Or do you have to do some conversion in this case. I've looked at the indices of the input mesh, and think both scenarios can appear.

**Reply by changkun** [reply]
Only the left is valid, otherwise, we get a flipped triangle (e.g. on the right, adjacency faces have flipped normals)

**Comment by Consti10** [comment]
Should we assign the member "vert" of each Halfedge the vertex the halfedge originated at or the Vertex the halfedge points to ?
I guess both would work (since you can can always use next.vert / prev.vert to get either of the two) but make the implementation(s) incompatible with each other.

**Reply by Consti10** [reply]
OpenMesh for example assigns each Halfedge the vertex it points to: [link](https://www.graphics.rwth-aachen.de/media/openmesh_static/Documentations/OpenMesh-5.2-Documentation/a00016.html)

**Reply by changkun** [reply]
I think a halfedge should points to the starting vertex rather than the ending vertex.

**Reply by changkun** [reply]
BTW: The reason is that calculating the vector will be must easier and faster, otherwise one has to travel every edge to get where is the starting vertex of the edge.

**Comment by nico778** [comment]
Hi, Question related to teaser.py. How are we supposed to create the quad bunny and the textured bunny? Converting the existing bunny doesn't really work for me, and I don't know where to start with the texture.

**Reply by changkun** [reply]
The quad meshed bunny only shows wireframe, hence it is possible to simply use the blender's built-in remesh feature to either generate a quad meshed bunny on the fly or in advance from the given triangle mesh. As showed in the class.

Texture is indeed a challenge since there is no texture given in the current mesh. Let's make it as an optional (sub-)task for rendering the teaser.

**Reply by nico778** [reply]
Ah okay, thanks, then I have to change my approach.

**Comment by MarlQ** [comment]
Is the deadline start or end of Monday?

**Reply by changkun** [reply]
The deadline is 01.11.2021 00:00:00, see https://github.com/mimuc/gp#homework-projects

**Reply by MarlQ** [reply]
That is the start of Monday I assume? (Meaning today is the last day)

**Reply by changkun** [reply]
Here is a countdown calculator: https://www.timeanddate.com/countdown/generic?p0=168&msg=01.11.2021+00%3A00%3A00&font=cursive

**Reply by nico778** [reply]
yes  today 23:59 I think

**Comment by JohannesMerkt** [comment]
Task 2 was optional right?

**Reply by changkun** [reply]
As discussed here (https://github.com/mimuc/gp/discussions/29#discussioncomment-1565868), the answer is yes. Here is a committed change (https://github.com/mimuc/gp/commit/6d98cceda29aa7e456944ed323e715b58f6c1b15).

**Comment by changkun** [comment, marked as answer]
Close due to the end of the submission period.
